<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>articull.utils package &#8212; articull 0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=27fed22d" />
    <script src="_static/documentation_options.js?v=2709fde1"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="articull.train package" href="articull.train.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="articull-utils-package">
<h1>articull.utils package<a class="headerlink" href="#articull-utils-package" title="Link to this heading">¶</a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading">¶</a></h2>
</section>
<section id="module-articull.utils.utils_bams">
<span id="articull-utils-utils-bams-module"></span><h2>articull.utils.utils_bams module<a class="headerlink" href="#module-articull.utils.utils_bams" title="Link to this heading">¶</a></h2>
<p>utils_bams.py</p>
<p>This module provides utility functions for working with BAM files and genomic variants.
It includes functionality for matching variants to BAM filenames, determining if reads are from high-quality normal cells,
opening SAM/BAM files, and generating reads at specific positions.</p>
<dl class="simple">
<dt>Functions:</dt><dd><dl class="simple">
<dt>match_variants_to_filenames(df, data_dirs):</dt><dd><p>Matches genomic variants to BAM filenames based on their chromosomal location.</p>
</dd>
<dt>is_normal(pileupread, labels=None):</dt><dd><p>Determines if a given pileup read is from a high-quality normal cell.</p>
</dd>
<dt>get_sam_path(path):</dt><dd><p>Opens a SAM/BAM file using pysam and returns the AlignmentFile object.</p>
</dd>
<dt>get_sam(data_dir, filename):</dt><dd><p>Opens a SAM/BAM file using pysam.AlignmentFile.</p>
</dd>
<dt>generate_reads(samfile, x, labels=None, min_base_quality=20, min_mapping_quality=0):</dt><dd><p>Generates reads at a specified position and reports whether they match the variant allele or not.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py function">
<dt class="sig sig-object py" id="articull.utils.utils_bams.generate_reads">
<span class="sig-prename descclassname"><span class="pre">articull.utils.utils_bams.</span></span><span class="sig-name descname"><span class="pre">generate_reads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">samfile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_base_quality</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_mapping_quality</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#articull.utils.utils_bams.generate_reads" title="Link to this definition">¶</a></dt>
<dd><p>Generates reads at a specified position and reports whether they match the variant allele or not.</p>
<dl class="field-list">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samfile</strong> (<em>pysam.AlignmentFile</em>) – The SAM/BAM file to read from.</p></li>
<li><p><strong>x</strong> (<em>dict</em>) – A dictionary containing the following keys:
- ‘chrm’ (str): Chromosome name.
- ‘pos’ (int): 1-based position of the variant.
- ‘ref_allele’ (str): Reference allele.
- ‘alt_allele’ (str): Alternate allele.
- ‘var_type’ (str): Variant type (‘SNP’, ‘DNP’, ‘DEL’, ‘INS’).</p></li>
<li><p><strong>labels</strong> (<em>dict</em><em>, </em><em>optional</em>) – A dictionary of labels for normal reads. Defaults to None.</p></li>
<li><p><strong>min_base_quality</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimum base quality for a read to be considered. Defaults to 20.</p></li>
<li><p><strong>min_mapping_quality</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimum mapping quality for a read to be considered. Defaults to 0.</p></li>
</ul>
</dd>
<dt class="field-even">Yields<span class="colon">:</span></dt>
<dd class="field-even"><p><em>tuple</em> –</p>
<dl class="simple">
<dt>A tuple containing:</dt><dd><ul class="simple">
<li><p>pileupread (pysam.PileupRead): The read from the pileup.</p></li>
<li><p>bool: True if the read matches the alternate allele, False otherwise.</p></li>
<li><p>bool: True if the read is normal according to the labels, False otherwise.</p></li>
</ul>
</dd>
</dl>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>Only reports reads with given alt or ref alleles and skips other alternate alleles.</p></li>
<li><p>There is processing for DEL, INS, and DNPs, but the code is not currently used. Realignment in mutect2
affects these a lot more, so typically there are very few actual supporting reads. This would be good
to revisit in the future.</p></li>
<li><p>It’d be nice to speed this step up, perhaps switching a portion of extract_features to directly use CLI
samtools. But I haven’t been able to get that to work in a straightforward way. Future extension?</p></li>
</ul>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="articull.utils.utils_bams.get_sam">
<span class="sig-prename descclassname"><span class="pre">articull.utils.utils_bams.</span></span><span class="sig-name descname"><span class="pre">get_sam</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#articull.utils.utils_bams.get_sam" title="Link to this definition">¶</a></dt>
<dd><p>Opens a SAM/BAM file using pysam.AlignmentFile.
This function attempts to open a SAM/BAM file with the given filename in the specified directory.
If the file is not found, it searches for a file with a suffix appended to the filename and tries to open it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_dir</strong> (<em>str</em>) – The directory where the SAM/BAM file is located.</p></li>
<li><p><strong>filename</strong> (<em>str</em>) – The name of the SAM/BAM file to open.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An object representing the opened SAM/BAM file.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pysam.AlignmentFile</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – If the file is not found in the specified directory.</p></li>
<li><p><strong>ValueError</strong> – If no file with the expected suffix is found in the directory.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="articull.utils.utils_bams.get_sam_path">
<span class="sig-prename descclassname"><span class="pre">articull.utils.utils_bams.</span></span><span class="sig-name descname"><span class="pre">get_sam_path</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#articull.utils.utils_bams.get_sam_path" title="Link to this definition">¶</a></dt>
<dd><p>Opens a SAM/BAM file using pysam and returns the AlignmentFile object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>path</strong> (<em>str</em>) – The file path to the SAM/BAM file.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An object representing the opened SAM/BAM file.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pysam.AlignmentFile</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>It should be checked if <cite>get_sam</cite> or <cite>get_sam_path</cite> is currently used and clean up the other.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="articull.utils.utils_bams.is_normal">
<span class="sig-prename descclassname"><span class="pre">articull.utils.utils_bams.</span></span><span class="sig-name descname"><span class="pre">is_normal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pileupread</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#articull.utils.utils_bams.is_normal" title="Link to this definition">¶</a></dt>
<dd><p>Determines if a given pileup read is from a high-quality normal cell.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pileupread</strong> (<em>pysam.PileupRead</em>) – The pileup read to be evaluated.</p></li>
<li><p><strong>labels</strong> (<em>pandas.DataFrame</em><em>, </em><em>optional</em>) – A DataFrame containing cell labels with
‘is_high_quality_normal_cell’ column. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the cell is a high-quality normal cell, False otherwise.
None: If labels are not provided.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="articull.utils.utils_bams.match_variants_to_filenames">
<span class="sig-prename descclassname"><span class="pre">articull.utils.utils_bams.</span></span><span class="sig-name descname"><span class="pre">match_variants_to_filenames</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_dirs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#articull.utils.utils_bams.match_variants_to_filenames" title="Link to this definition">¶</a></dt>
<dd><p>Matches genomic variants to BAM filenames based on their chromosomal location.
This function assumes that BAM files are merged and in region format. For each variant
in the provided DataFrame, it matches the variant to the appropriate BAM file that
contains the corresponding locus.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – A DataFrame containing genomic variants with columns ‘chrm’ (chromosome)
and ‘pos’ (position).</p></li>
<li><p><strong>data_dirs</strong> (<em>list</em>) – A list of directories where BAM files are stored. Only the first directory
in the list is used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>The input DataFrame with an additional column ‘filename’ that contains the</dt><dd><p>matched BAM filename for each variant. If no matching file is found, the filename will be None.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function may not be used anymore since switching to merged BAMs, but it should be checked.</p>
</div>
</dd></dl>

</section>
<section id="module-articull.utils.utils_io">
<span id="articull-utils-utils-io-module"></span><h2>articull.utils.utils_io module<a class="headerlink" href="#module-articull.utils.utils_io" title="Link to this heading">¶</a></h2>
<p>utils_io.py</p>
<p>This module provides utility functions for reading and processing variant data from MAF and VCF files.
It includes functionality for extracting variant information and reading VCF files.</p>
<dl class="simple">
<dt>Functions:</dt><dd><dl class="simple">
<dt>get_variants(filename, file_type=”maf”):</dt><dd><p>Extracts variant information from a given file in MAF or VCF format.</p>
</dd>
<dt>read_vcf(path):</dt><dd><p>Reads a VCF (Variant Call Format) file, optionally compressed with gzip, and returns a pandas DataFrame.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py function">
<dt class="sig sig-object py" id="articull.utils.utils_io.get_variants">
<span class="sig-prename descclassname"><span class="pre">articull.utils.utils_io.</span></span><span class="sig-name descname"><span class="pre">get_variants</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'maf'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#articull.utils.utils_io.get_variants" title="Link to this definition">¶</a></dt>
<dd><p>Extracts variant information from a given file in MAF or VCF format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str</em>) – The path to the input file. The file can be in MAF (.maf or .maf.gz) or VCF (.vcf or .vcf.gz) format.</p></li>
<li><p><strong>file_type</strong> (<em>str</em><em>, </em><em>optional</em>) – The type of the input file. Default is “maf”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A DataFrame containing the following columns:</dt><dd><ul class="simple">
<li><p>’chrm’: Chromosome</p></li>
<li><p>’pos’: Position</p></li>
<li><p>’ref_allele’: Reference allele</p></li>
<li><p>’alt_allele’: Alternate allele</p></li>
<li><p>’var_type’: Variant type (only ‘SNP’ variants are included)</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>Exception</strong> – If the input file type is not MAF or VCF.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="articull.utils.utils_io.read_vcf">
<span class="sig-prename descclassname"><span class="pre">articull.utils.utils_io.</span></span><span class="sig-name descname"><span class="pre">read_vcf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#articull.utils.utils_io.read_vcf" title="Link to this definition">¶</a></dt>
<dd><p>Reads a VCF (Variant Call Format) file, optionally compressed with gzip, and returns a pandas DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>path</strong> (<em>str</em>) – The file path to the VCF file. The file can be plain text or gzip compressed.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the VCF data with columns [‘CHROM’, ‘POS’, ‘ID’, ‘REF’, ‘ALT’, ‘QUAL’, ‘FILTER’, ‘INFO’].</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-articull.utils.utils_setup">
<span id="articull-utils-utils-setup-module"></span><h2>articull.utils.utils_setup module<a class="headerlink" href="#module-articull.utils.utils_setup" title="Link to this heading">¶</a></h2>
<p>utils_setup.py</p>
<p>This module provides utility functions for setting up the computational environment.
It includes functionality for setting the number of cores for OpenBLAS and MKL libraries,
and initializing the pandarallel library for parallel processing with pandas.</p>
<dl class="simple">
<dt>Functions:</dt><dd><dl class="simple">
<dt>setup_ncores(ncores=None):</dt><dd><p>Sets up the number of cores to be used by OpenBLAS and MKL libraries.</p>
</dd>
<dt>setup_pandarallel(mode, ncores=None):</dt><dd><p>Initializes the pandarallel library with the specified mode and number of cores.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py function">
<dt class="sig sig-object py" id="articull.utils.utils_setup.setup_ncores">
<span class="sig-prename descclassname"><span class="pre">articull.utils.utils_setup.</span></span><span class="sig-name descname"><span class="pre">setup_ncores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ncores</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#articull.utils.utils_setup.setup_ncores" title="Link to this definition">¶</a></dt>
<dd><p>Sets up the number of cores to be used by OpenBLAS and MKL libraries.
This function sets the environment variables ‘OPENBLAS_NUM_THREADS’ and ‘MKL_NUM_THREADS’
to control the number of threads used by these libraries. This is primarily done to limit
the number of cores used by numpy, which otherwise uses as many cores as it wants.
Be a good citizen and don’t hog all the cores on the cluster.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function must be called before numpy is imported, or else numpy will use as many
cores as it wants.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ncores</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of cores to be used. If not provided, the default
number of cores will be used.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="articull.utils.utils_setup.setup_pandarallel">
<span class="sig-prename descclassname"><span class="pre">articull.utils.utils_setup.</span></span><span class="sig-name descname"><span class="pre">setup_pandarallel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mode</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncores</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#articull.utils.utils_setup.setup_pandarallel" title="Link to this definition">¶</a></dt>
<dd><p>Initializes the pandarallel library with the specified mode and number of cores.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mode</strong> (<em>str</em>) – The mode in which to run pandarallel. If mode is ‘classify’, the progress bar will be disabled.</p></li>
<li><p><strong>ncores</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of cores to use for parallel processing.
If not specified, pandarallel will use the default number of cores.</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The progress bar is best used when output is running to the command line.
When it’s being output to a log file, it may result in an excessive number of lines.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-articull.utils">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-articull.utils" title="Link to this heading">¶</a></h2>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">articull</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">src</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="articull.html">articull package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="articull.html#subpackages">Subpackages</a></li>
<li class="toctree-l3"><a class="reference internal" href="articull.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="articull.html#module-articull.main">articull.main module</a></li>
<li class="toctree-l3"><a class="reference internal" href="articull.html#module-articull">Module contents</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="generated/articull.html">articull</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="modules.html">src</a><ul>
  <li><a href="articull.html">articull package</a><ul>
      <li>Previous: <a href="articull.train.html" title="previous chapter">articull.train package</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2025, Gryte Satas.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.2.1</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="_sources/articull.utils.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>